### ISSUE TITLE
Provide a mechanism to list all the global variables

### ISSUE THREAD

**Issue by soywiz** [JetBrains/kotlin-native#1781]
Would be nice if idea, or gradle allows to show al edit-time or compile-time all the global properties with backing field of a project. Right now it is hard to find them in large codebases since they are not explicitly annotated with the storage type (frozen or threadlocal).

Also would be nice if the mutation exceptions shows or tries to guess the root global that references the mutation object and produces the frozen.
Right now I have a case that i was not able to find. Fixed all the obvios places and still happens, so im a bit puzzled and displaying the root or ideally one reference tree from the root to the instance tried to mutate, would help a lot even if it is expensive to compute since in a good program wont happen.

This two features would make much easier to deal with the frozen globals.

**Comment by olonho** [comment]
Well, what we could implement is a runtime operation or annotation on Kotlin object, which throws an exception if someone tries to freeze this object. Would that help you?

**Comment by soywiz** [comment]
```
@MustNotFreeze
class MyClass1
```
or similar?

That would help. Would it show the global root that produced the freeze? That would be fantastic (to avoid having to go down manually). Each test takes 15 min in my case.

---

The problem is that I have a pretty big codebase with a lot of objects and companion objects, but also global variables or "constants", most of the objects only provide plain methods but some of them have state, and some others I thought it only had constants, but ended having unintentional state after some time of coding, but now it is hard to find exactly where. I have only one single thread so had no concurrency problems before. Most of the global objects only mutated the state for caching purposes.

Also being a runtime error, each time I want to test, I have to recompile a library that takes 12 minutes, plus 2,5 minutes for the main executable. It requires a lot of patience so possible hints before that would help a lot.

```
var globalVariable = MyClass3()
object MyObject{
   fun test() {
   }
   val objectConstant = MyClass2()
}

val anotherGlobalConstant = MyClass1()
fun myfunctionUsingAnotherGlobalConstant() {
}
```

I sometimes have global variables/constants before some functions. They are not inside objects, but still they are global.
I cannot mark them/evaluate if they require ThreadLocal if I don't know they are there.

I guess the compiler could check at compile-time if a simple object tree is effectively immutable if only have vals to other immutable classes and could be discarded for a potential compile-time/edit-time report.

**Comment by olonho** [comment]
Please try https://github.com/JetBrains/kotlin-native/pull/1783 and report, how useful it is. Idea is you call ensureNeverFrozen() and then whenever freeze reaches object marked this way - it throws an exception.

**Comment by soywiz** [comment]
I will try later today 👍

**Comment by soywiz** [comment]
I was going to suggest to change to `fun <T> T.ensureNeverFrozen(): T = this.apply { ensureNeverFrozen() }`, but in the end I had to use it in an `init {}` block at the start of the class.

https://github.com/korlibs/korlibs/tree/freeze.test
https://github.com/korlibs/korlibs/commit/4101a9e153c58ffa0992e2bca5279a63d3030752

Not sure if it s a bug, or something I don't understand. I get this exception:

```
Uncaught exception from Kotlin's main: konan.worker.FreezingException: freezing of function <anonymous> (Kotlin reflection is not available) has failed, first blocker is com.soywiz.korui.MacosNativeEventLoop@97f55bd8
        at 3   sample1-native.kexe                 0x000000010992c986 kfun:kotlin.Exception.<init>(kotlin.String?)kotlin.Exception + 70
        at 4   sample1-native.kexe                 0x000000010992c8a6 kfun:kotlin.RuntimeException.<init>(kotlin.String?)kotlin.RuntimeException + 70
        at 5   sample1-native.kexe                 0x00000001099da356 kfun:konan.worker.FreezingException.<init>(kotlin.Any;kotlin.Any)konan.worker.FreezingException + 230
        at 6   sample1-native.kexe                 0x00000001099da55a ThrowFreezingException + 90
        at 7   sample1-native.kexe                 0x00000001099ee390 FreezeSubgraph + 304
        at 8   sample1-native.kexe                 0x00000001099659d7 kfun:konan.worker.freeze@#GENERIC.()Generic + 55
        at 9   sample1-native.kexe                 0x00000001099d9948 kfun:konan.worker.AtomicLazyImpl.<init>(kotlin.Function0<#GENERIC>)konan.worker.AtomicLazyImpl<#GENERIC> + 88
        at 10  sample1-native.kexe                 0x00000001099da21a kfun:konan.worker.atomicLazy(kotlin.Function0<#GENERIC>)Generic + 90
        at 11  sample1-native.kexe                 0x0000000109bbf7f0 kfun:com.soywiz.korui.MacosNativeEventLoop.<init>()com.soywiz.korui.MacosNativeEventLoop + 176
        at 12  sample1-native.kexe                 0x0000000109bbf6d3 kfun:com.soywiz.korui.KoruiEventLoop.create()com.soywiz.korio.async.EventLoop + 83
        at 13  sample1-native.kexe                 0x0000000109c0489d kfun:com.soywiz.korge.Korge.invoke$default(com.soywiz.korge.scene.Module;kotlin.Array<kotlin.String>;com.soywiz.korag.AGContainer?;com.soywiz.korui.event.EventDispatcher;kotlin.reflect.KClass<out_com.soywiz.korge.scene.Scene>;kotlin.collections.List<kotlin.Any>;com.soywiz.klock.TimeProvider;com.soywiz.korinject.AsyncInjector;kotlin.Boolean;kotlin.Boolean;kotlin.Function1<com.soywiz.korge.view.Views,kotlin.Unit>;com.soywiz.korio.async.EventLoop;kotlin.Int) + 1005
        at 14  sample1-native.kexe                 0x000000010992a71a kfun:example.main(kotlin.Array<kotlin.String>) + 1162
        at 15  sample1-native.kexe                 0x000000010992a252 kfun:main(kotlin.Array<kotlin.String>) + 50
        at 16  sample1-native.kexe                 0x000000010992a216 EntryPointSelector + 6
        at 17  sample1-native.kexe                 0x000000010992a101 Konan_run_start + 113
        at 18  sample1-native.kexe                 0x000000010992a07b Konan_main + 27
        at 19  libdyld.dylib                       0x00007fff5d929015 start + 1
        at 20  ???                                 0x0000000000000001 0x0 + 1
```

```
actual object KoruiEventLoop {
	actual fun create(): EventLoop = MacosNativeEventLoop().apply { ensureNeverFrozen() } // this ensureNeverFrozen did nothing.
}

open class MacosNativeEventLoop : EventLoop() {
	init {
		ensureNeverFrozen() // This one worked
	}
	var app: NSApplication? by atomicRef<NSApplication?>(null)

	val ag: AG by atomicLazy { AGOpenglFactory.create(this).create(this) }
	var listener = object : KMLWindowListener() {
	}

class atomicRef<T>(initial: T) {
	val value = NewAtomicReference<T?>(initial)

	inline operator fun getValue(obj: Any?, property: KProperty<*>): T {
		@Suppress("UNCHECKED_CAST")
		return this.value.get() as T
	}

	inline operator fun setValue(obj: Any?, property: KProperty<*>, v: T) {
		this.value.set(v)
	}
}
```

As far as I know the `KoruiEventLoop.create()` is called from inside a plain method after main (no global) and it is as much stored in the coroutineContext. So not sure what's going on. Any tips?

**Comment by olonho** [comment]
Error message says that anonymous lambda captures MacosNativeEventLoop, which is indeed the case (you refer to `this` in atomicLazy's lambda). What are you trying to achieve here, actually?

**Comment by soywiz** [comment]
Ok. I didn't know that atomicLazy was manually freezing the initalizer:
https://github.com/JetBrains/kotlin-native/blob/7e517ef6f1154a59473dd6b8a4c3dd39fe5c1360/runtime/src/main/kotlin/konan/worker/Atomics.kt#L160

Which probably makes sense, though limites its uses.

I still don't know why isn't that class freeze-friendly because AG doesn't have mutable state, the listener is not mutated, and the app is an atomicRef to something external to kotlin handled by objective-c.
Is it because when the atomicLazy is being constructed, the MacosNativeEventLoop is still constructing and mutates after that while constructing? If that's the case, it is a bit complicated to understand in the first instance. And could the initializer freeze be delayed to here https://github.com/JetBrains/kotlin-native/blob/7e517ef6f1154a59473dd6b8a4c3dd39fe5c1360/runtime/src/main/kotlin/konan/worker/Atomics.kt#L167 ?

I came up with that code because initially I had a global reference to the EventLoop, and I tried to make that class freeze-friendly, by introducing that. Then I found the global reference, removed it, but the freeze still happened, so I was puzzled.

Knowing that and with `ensureNeverFrozen` maybe now I can find offending places a bit easier.
It would help also, if it is possible, to include the file and line of anonymous lambdas as part of the debug information.
